src/app/Components/shopping-list/shopping-list.component.ts

```diff
@@ -0,0 +1,78 @@
+import { Component, OnInit } from '@angular/core';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { Router } from '@angular/router';
+import { ShopListService } from 'src/app/Services/shop-list.service';
+
+@Component({
+  selector: 'app-shopping-list',
+  templateUrl: './shopping-list.component.html',
+  styleUrls: ['./shopping-list.component.scss'],
+})
+export class ShoppingListComponent implements OnInit {
+  ingredients!: any[];
+  modeUse = false;
+  modeClearBtn = false;
+  shoppingForm!: FormGroup;
+
+  constructor(private dataShopList: ShopListService, private router: Router) {}
+
+  ngOnInit(): void {
+    this.ingredients = this.dataShopList.ingredients;
+    this.shoppingForm = new FormGroup({
+      name: new FormControl('', [Validators.required]),
+      amount: new FormControl(0, [Validators.required, Validators.min(1)]),
+    });
+    this.shoppingForm.valueChanges.subscribe((res) => {
+      if (res.amount > 0) {
+        this.modeClearBtn = true;
+      } else {
+        this.modeClearBtn = false;
+      }
+    });
+  }
+
+  addIngredient() {
+    if (this.shoppingForm.invalid) {
+      return;
+    }
+
+    const newId = this.dataShopList.ingredients.length + 1;
+    const newIngredient = { id: newId, ...this.shoppingForm.value };
+
+    this.dataShopList.addIngredient(newIngredient);
+    this.shoppingForm.reset({ name: '', amount: 0 });
+  }
+
+  onIngredientDetail(ingredient: any) {
+    this.modeUse = true;
+    this.shoppingForm.get('name')?.setValue(ingredient.name);
+    this.shoppingForm.get('amount')?.setValue(ingredient.amount);
+  }
+
+  updateIngredient() {
+    if (this.shoppingForm.invalid) {
+      return;
+    }
+
+    const id = +this.router.url.split('/')[2];
```

Comment:
this is a practice assignment so u can get the `id` from the URL, but in a real app, u can only trust the init data (the data u get from the start of the component).
try to create a new `FormControl` for the `id` field when opening an ingredient detail so we can manage its data better.